File: packages/python/elizaos/features/core_capabilities/secrets/actions/manage_secret.py
"""Manage secret action.

Provides list, delete, and status operations on secrets through a
natural-language interface.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.types import Action, ActionResult, Content

from ..types import SecretContext, SecretLevel

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

    from ..service import SecretsService


@dataclass
class ManageSecretAction:
    name: str = "MANAGE_SECRET"
    similes: list[str] = field(
        default_factory=lambda: ["DELETE_SECRET", "REMOVE_SECRET", "LIST_SECRETS", "SECRET_STATUS"]
    )
    description: str = "List, delete, or check the status of stored secrets"

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        return True

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        secrets_svc: SecretsService | None = None
        for svc in (runtime.services or {}).values():
            if getattr(svc, "service_type", None) == "secrets":
                secrets_svc = svc  # type: ignore[assignment]
                break

        if secrets_svc is None:
            return ActionResult(text="Secrets service is not available.", success=False)

        text = (message.content.text if message.content else "").strip().lower()
        context = SecretContext(level=SecretLevel.GLOBAL, agent_id=str(runtime.agent_id))

        # Determine operation
        if "delete" in text or "remove" in text:
            # Extract key to delete
            words = text.split()
            key = next(
                (w.upper() for w in words if w.upper() != w.lower() or "_" in w),
                None,
            )
            if not key:
                msg = "Please specify which secret to delete."
                if callback:
                    await callback(Content(text=msg, actions=["MANAGE_SECRET"]))
                return ActionResult(text=msg, success=False)

            key = key.upper()
            success = await secrets_svc.delete(key, context)
            msg = (
                f"Secret '{key}' deleted successfully." if success else f"Secret '{key}' not found."
            )
            if callback:
                await callback(Content(text=msg, actions=["MANAGE_SECRET"]))
            return ActionResult(text=msg, values={"success": success}, success=success)

        # Default: list secrets
        metadata = await secrets_svc.list(context)
        if not metadata:
            msg = "No secrets are currently stored."
        else:
            lines = ["# Stored Secrets"]
            for key, config in metadata.items():
                status = config.status.value if hasattr(config.status, "value") else config.status
                lines.append(f"- {key}: {status} (level={config.level.value})")
            msg = "\n".join(lines)

        if callback:
            await callback(Content(text=msg, actions=["MANAGE_SECRET"]))

        return ActionResult(
            text=msg,
            values={"success": True, "secretCount": len(metadata)},
            data={"actionName": "MANAGE_SECRET", "keys": list(metadata.keys())},
            success=True,
        )

    @property
    def examples(self) -> list:
        return []


_inst = ManageSecretAction()

manage_secret_action = Action(
    name=_inst.name,
    similes=_inst.similes,
    description=_inst.description,
    validate=_inst.validate,
    handler=_inst.handler,
    examples=_inst.examples,
)

File: packages/typescript/src/features/secrets/actions/manage-secret.ts
/**
 * Manage Secret Action
 *
 * Comprehensive action for managing secrets through natural language.
 * Supports get, set, delete, and list operations at different levels.
 */

import { logger } from "../../../logger.ts";
import { extractSecretOperationTemplate as extractOperationTemplate } from "../../../prompts.ts";
import {
	type Action,
	type ActionExample,
	ChannelType,
	composePromptFromState,
	type HandlerCallback,
	type HandlerOptions,
	type IAgentRuntime,
	type JsonValue,
	type Memory,
	ModelType,
	type State,
} from "../../../types/index.ts";
import {
	SECRETS_SERVICE_TYPE,
	type SecretsService,
} from "../services/secrets.ts";
import type { SecretContext, SecretType } from "../types.ts";

/**
 * Type for secret management operation
 */
interface SecretOperation {
	operation: "get" | "set" | "delete" | "list" | "check";
	key?: string;
	value?: string;
	level?: "global" | "world" | "user";
	description?: string;
	type?: "api_key" | "secret" | "credential" | "url" | "config";
}

/**
 * Manage Secret Action
 */
export const manageSecretAction: Action = {
	name: "MANAGE_SECRET",
	similes: [
		"SECRET_MANAGEMENT",
		"HANDLE_SECRET",
		"SECRET_OPERATION",
		"GET_SECRET",
		"DELETE_SECRET",
		"LIST_SECRETS",
		"CHECK_SECRET",
	],
	description:
		"Manage secrets - get, set, delete, or list secrets at various levels",

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
		_options?: HandlerOptions,
	): Promise<boolean> => {
		const text = message.content.text?.toLowerCase() ?? "";
		const patterns = [
			/\b(get|show|what|retrieve)\b.*\b(secret|key|token|credential)/i,
			/\b(delete|remove|clear)\b.*\b(secret|key|token|credential)/i,
			/\b(list|show)\b.*\b(secrets|keys|tokens|credentials)/i,
			/\bdo i have\b.*\b(secret|key|token)/i,
			/\b(check|is)\b.*\b(secret|key|token)\b.*\b(set|configured)/i,
			/\bmy secrets\b/i,
			/\bwhat secrets\b/i,
		];
		if (!patterns.some((pattern) => pattern.test(text))) {
			return false;
		}

		return runtime.getService<SecretsService>(SECRETS_SERVICE_TYPE) !== null;
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
	) => {
		logger.info("[ManageSecret] Processing secret management request");

		// Security: Refuse to manage secrets in non-DM channels
		const channelType = message.content.channelType;
		if (channelType !== undefined && channelType !== ChannelType.DM) {
			logger.warn(
				"[ManageSecret] Refused: attempted to manage secrets in non-DM channel",
			);
			if (callback) {
				await callback({
					text: "I can't manage secrets in a public channel. Please send me a direct message (DM) for secret operations. Never share sensitive information in public channels.",
					action: "MANAGE_SECRET",
				});
			}
			return {
				success: false,
				text: "Refused: secrets can only be managed in DMs",
			};
		}

		const secretsService =
			runtime.getService<SecretsService>(SECRETS_SERVICE_TYPE);
		if (!secretsService) {
			if (callback) {
				await callback({
					text: "Secret management is not available.",
					action: "MANAGE_SECRET",
				});
			}
			return { success: false, text: "Secrets service not available" };
		}

		// Build state for prompt
		const currentState = state ?? (await runtime.composeState(message));

		// Extract operation from user message
		let operation: SecretOperation;
		try {
			const prompt = composePromptFromState({
				state: currentState,
				template: extractOperationTemplate,
			});

			const result = (await runtime.useModel(ModelType.OBJECT_SMALL, {
				prompt,
			})) as Record<string, JsonValue>;

			// Transform and validate result
			operation = {
				operation: (result.operation as SecretOperation["operation"]) || "list",
				key: result.key ? String(result.key) : undefined,
				value: result.value ? String(result.value) : undefined,
				level: result.level as SecretOperation["level"],
				description: result.description
					? String(result.description)
					: undefined,
				type: result.type as SecretOperation["type"],
			};
		} catch (error) {
			const errorMessage =
				error instanceof Error ? error.message : String(error);
			logger.error(
				`[ManageSecret] Failed to extract operation: ${errorMessage}`,
			);
			if (callback) {
				await callback({
					text: "I had trouble understanding what you want to do with secrets. Could you be more specific?",
					action: "MANAGE_SECRET",
				});
			}
			return {
				success: false,
				text: "Failed to extract operation from message",
			};
		}

		// Determine storage context
		const level = operation.level ?? "global";
		const context: SecretContext = {
			level,
			agentId: runtime.agentId,
			worldId: level === "world" ? (message.roomId as string) : undefined,
			userId: level === "user" ? (message.entityId as string) : undefined,
			requesterId: message.entityId as string,
		};

		// Execute the operation
		let responseText: string;

		switch (operation.operation) {
			case "get": {
				if (!operation.key) {
					responseText = "Please specify which secret you want to retrieve.";
					break;
				}

				const key = operation.key.toUpperCase().replace(/[^A-Z0-9_]/g, "_");
				const value = await secretsService.get(key, context);

				if (value) {
					// Never reveal full secret values - show partial
					const maskedValue = maskSecretValue(value);
					responseText = `Your ${key} is set to: ${maskedValue}`;
				} else {
					responseText = `I don't have a ${key} stored. Would you like to set one?`;
				}
				break;
			}

			case "set": {
				if (!operation.key || !operation.value) {
					responseText = "Please provide both a key and value to set a secret.";
					break;
				}

				const key = operation.key.toUpperCase().replace(/[^A-Z0-9_]/g, "_");

				try {
					const success = await secretsService.set(
						key,
						operation.value,
						context,
						{
							type: (operation.type as SecretType) ?? "secret",
							description: operation.description ?? "Set via conversation",
							encrypted: true,
						},
					);

					if (success) {
						responseText = `I've securely stored your ${key}.`;
					} else {
						responseText = `Failed to store ${key}. Please try again.`;
					}
				} catch (error) {
					responseText = `Error storing ${key}: ${error instanceof Error ? error.message : "Unknown error"}`;
				}
				break;
			}

			case "delete": {
				if (!operation.key) {
					responseText = "Please specify which secret you want to delete.";
					break;
				}

				const key = operation.key.toUpperCase().replace(/[^A-Z0-9_]/g, "_");
				const deleted = await secretsService.delete(key, context);

				if (deleted) {
					responseText = `I've deleted your ${key}.`;
				} else {
					responseText = `I couldn't find a ${key} to delete.`;
				}
				break;
			}

			case "list": {
				const metadata = await secretsService.list(context);
				const keys = Object.keys(metadata);

				if (keys.length === 0) {
					responseText = `You don't have any ${level} secrets stored yet.`;
				} else {
					const secretList = keys
						.map((key) => {
							const config = metadata[key];
							const status = config.status === "valid" ? "✓" : "⚠";
							return `• ${key} ${status}`;
						})
						.join("\n");

					responseText = `Here are your ${level} secrets:\n${secretList}`;
				}
				break;
			}

			case "check": {
				if (!operation.key) {
					responseText = "Please specify which secret you want to check.";
					break;
				}

				const key = operation.key.toUpperCase().replace(/[^A-Z0-9_]/g, "_");
				const exists = await secretsService.exists(key, context);

				if (exists) {
					const config = await secretsService.getConfig(key, context);
					const status =
						config?.status === "valid"
							? "valid"
							: (config?.status ?? "unknown");
					responseText = `Yes, ${key} is set and its status is: ${status}.`;
				} else {
					responseText = `No, ${key} is not set. Would you like to configure it?`;
				}
				break;
			}

			default:
				responseText =
					"I'm not sure what operation you want to perform. You can get, set, delete, list, or check secrets.";
		}

		if (callback) {
			await callback({
				text: responseText,
				action: "MANAGE_SECRET",
			});
		}

		return { success: true, text: responseText };
	},

	examples: [
		[
			{
				name: "{{user1}}",
				content: { text: "What secrets do I have?" },
			},
			{
				name: "{{agent}}",
				content: {
					text: "Here are your global secrets:\n• OPENAI_API_KEY ✓\n• ANTHROPIC_API_KEY ✓",
					action: "MANAGE_SECRET",
				},
			},
		],
		[
			{
				name: "{{user1}}",
				content: { text: "Do I have a Discord token set?" },
			},
			{
				name: "{{agent}}",
				content: {
					text: "No, DISCORD_BOT_TOKEN is not set. Would you like to configure it?",
					action: "MANAGE_SECRET",
				},
			},
		],
		[
			{
				name: "{{user1}}",
				content: { text: "Delete my old Twitter API key" },
			},
			{
				name: "{{agent}}",
				content: {
					text: "I've deleted your TWITTER_API_KEY.",
					action: "MANAGE_SECRET",
				},
			},
		],
	] as ActionExample[][],
};

/**
 * Mask a secret value for display
 */
function maskSecretValue(value: string): string {
	if (value.length <= 8) {
		return "****";
	}

	const visibleStart = value.slice(0, 4);
	const visibleEnd = value.slice(-4);
	const maskedLength = Math.min(value.length - 8, 20);
	const mask = "*".repeat(maskedLength);

	return `${visibleStart}${mask}${visibleEnd}`;
}
